source/slang/check.cpp

```diff
@@ -9535,6 +9581,28 @@ namespace Slang
             if(isGlobalShaderParameter(varDeclRef.getDecl()))
```

Comment:
Working out if is LValue might be better split out as a function so as once it's determined it's isn't the other tests aren't needed.

Reply:
Agreed. The handling of l-value-ness in the front-end isn't especially great anyway, so it might be worth us taking the time to think about how we want to approach it systematically. The current approach introduces the `QualType` idea as a way to represent a `Type*` plus a bit for l-value-ness, but having everything need to store/handle `QualType` feels like a lot of overhead for what amounts to a very local check for whether something is writable (aside: we could technically hide the l-value-ness bit in the low bit of a `Type*` to avoid any storage overhead, but our pervasive use of `RefPtr` makes tricks like that harder).

An very different alternative would be to have a general-purpose `isWritable(Expr*)` that inspects the structure and/or type of an expression to figure out whether it should be considered an l-value and then only compute this in the exact place(s) needed (and not store it thereafter).